common/flatpak-appdata.c

```diff
@@ -21,7 +21,168 @@
+  data->text = g_string_new ("");
+
+  return data;
+}
+
+static void
+start_element (GMarkupParseContext *context,
+               const char *element_name,
+               const char **attribute_names,
+               const char **attribute_values,
+               gpointer user_data,
+               GError **error)
+{
+  ParserData *data = user_data;
+
+  g_assert (data->text->len == 0);
+  g_assert (data->lang == NULL);
+
+  if (g_str_equal (element_name, "id"))
+    {
+      data->in_text = TRUE;
+    }
+  else if (g_str_equal (element_name, "name") ||
+           g_str_equal (element_name, "summary"))
+    {
+      const char *lang = NULL;
+      if (data->in_component &&
+          g_markup_collect_attributes (element_name,
+                                       attribute_names,
+                                       attribute_values,
+                                       error,
+                                       G_MARKUP_COLLECT_STRING | G_MARKUP_COLLECT_OPTIONAL, "xml:lang", &lang,
+                                       G_MARKUP_COLLECT_INVALID))
+        {
+          if (lang)
+            data->lang = g_strdup (lang);
+          else
+            data->lang = g_strdup ("en");
+          data->in_text = TRUE;
+        }
+    }
+  else if (g_str_equal (element_name, "project_license"))
+    {
+      if (data->in_component)
+        data->in_text = TRUE;
+    }
+  else if (g_str_equal (element_name, "release"))
+    {
+      const char *timestamp;
+      const char *version;
+
+      if (data->in_component &&
+          g_markup_collect_attributes (element_name,
+                                       attribute_names,
+                                       attribute_values,
+                                       error,
+                                       G_MARKUP_COLLECT_STRING, "timestamp", &timestamp,
+                                       G_MARKUP_COLLECT_STRING, "version", &version,
+                                       G_MARKUP_COLLECT_INVALID))
+        {
+          guint64 ts = g_ascii_strtoull (timestamp, NULL, 10);
+          if (ts > data->timestamp)
+            {
+              data->timestamp = ts;
+              g_free (data->version);
+              data->version = g_strdup (version);
+            }
+        }
+    }
+}
+
+static void
+end_element (GMarkupParseContext *context,
+             const char *element_name,
+             gpointer user_data,
+             GError **error)
+{
+  ParserData *data = user_data;
+  g_autofree char *text = NULL;
+
+  if (data->in_text)
+    {
+      text = g_strdup (data->text->str);
+      g_string_truncate (data->text, 0);
+      data->in_text = FALSE;
+    }
+
+  if (g_str_equal (element_name, "component"))
+    {
+      data->in_component = FALSE;
+    }
+  else if (g_str_equal (element_name, "id"))
+    {
+      if (g_str_equal (text, data->app_id))
+        data->in_component = TRUE;
```

Comment:
This seems to rely on "id" being the first (or at least early) element in the component. Maybe a better approach is to always collect everything, including the id, and then return in end of componente if it matches, else clear everything.

Also, we need to see if id == app_id || id == app_id + ".desktop" for backwards compat.

Reply:
Yeah, thats true